drivers: flash_mspi_nor: Apply a few improvements

anangl opened the pull request:
Apply a few improvements in the flash_mspi_nor driver, most notably:
- use parameters from SFDP for all the used flash commands with possibility to override those through dts (with the `read-command`, `write-command`, and `rx-dummy` properties)
- use all available erase types as specified by SFDP
- allow using all IO modes
- add support for switching to 4-byte addressing mode
- complete handling of Quad Enable Requirements and add handling of Octal Enable Requirements
- add Soft Reset.

See the individual commit messages for details.

anangl committed:
drivers: flash_mspi_nor: Get info from dts SFDP arrays

Get parameters for used flash commands and requirements for enabling
Quad and Octal modes from dts uint8-arrays containing data read from
SFDP tables for particular flash chips.
Also introduce `pre_init` quirk that allows alteration of the above
parameters or complementation of them in a specific way for particular
flash chip families.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: flash_mspi_nor: Refactor handling of commands

- Use standard operation codes and parameters from SFDP for handling
  the used flash commands (allow to override some of them through dts
  with the `read-command`, `write-command`, and `rx-dummy` properties)
- Use all available erase types as specified by SFDP
- Allow using all IO modes
- Add support for switching to 4-byte addressing mode
- Use common functions for reading and writing of status registers
  and for enabling write operations
- Switch IO mode (between the target one and Single IO) in a common
  function that performs transfers and do it only when required for
  a given command
- Make checking of JEDEC ID at initialization optional

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: flash_mspi_nor: Complete handling of QER, add handling of OER

Complete implementation of quad_enable_set() by adding support for all
possible Quad Enable Requirements (QER) as specified by the SFDP JEDEC
standard (JESD216). Add also corresponding octal_enable_set() to handle
Octal Enable Requirements.

Also remove initial waiting from mxicy_mx25r_post_switch_mode() which
became unneeded, as now such waiting is done in cmd_wrsr() which is
called at the end of quad_enable_set().

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: flash_mspi_nor: Add Soft Reset

Add implementation of the most common Soft Reset routine (sequence of
reset enable instruction 0x66 and reset instruction 0x99).

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: flash_mspi_nor: Remove undesirable initial Quad disabling

Using a GPIO reset for a flash chip that has a dual function pin
(RESET# or SIO3) and is to be used in Quad mode is rather a bad idea
and so is clearing of the Quad Enable bit at every initialization
of the flash driver, since this bit is usually non-volatile, so such
operation means unnecessary wearing of the flash chip. Soft Reset
should be use instead in such case.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

danieldegrasse reviewed (commented):
I'm not specifically opposed to using SFDP for this driver, but I honestly doubt we are going to get the best performance out of most flash devices with this method. My primary concern is that not all flash devices actually implement SFDP discoverability, particularly for octal or quad settings (as least as far as I've seen)

Do you have any thoughts on the approach I took in https://github.com/zephyrproject-rtos/zephyr/pull/88920? I think these methods can coexist if they need to, but I personally feel like we may just be better off using compatibles for flash settings at build time, and JEDEC ID to select flash settings at runtime.

swift-tk reviewed (commented):
This reminds me of the attempt I took adding mspi_nor... The global storage needed by SFDP would explode flash_data as we continue to add more DWORDS.
Therefore, IMO, it would better the SFDP can have its own data struct and separate .c file and then we could have it enabled through Kconfig.

swift-tk commented:
> I'm not specifically opposed to using SFDP for this driver, but I honestly doubt we are going to get the best performance out of most flash devices with this method. My primary concern is that not all flash devices actually implement SFDP discoverability, particularly for octal or quad settings (as least as far as I've seen)
> 
> Do you have any thoughts on the approach I took in #88920? I think these methods can coexist if they need to, but I personally feel like we may just be better off using compatibles for flash settings at build time, and JEDEC ID to select flash settings at runtime.

The SFDP should be used for generic flash that does support it and enable through a Kconfig and in combination of a dts property. With SFDP, it should load cmd and other properties through DWORDS instead of using the generic or the vendor table. It is essentially a different way of runtime probing, so I think it could coexist well provided that it doesn't get in and mess up the core all over the place.

anangl commented:
@danieldegrasse @swift-tk I moved almost all processing of SFDP to a separate file (`FLASH_SIZE()` and `FLASH_PAGE_EXP()` macros are the only exceptions). SFDP processing is done only at build time, using data from dts arrays. Obtained information can be overridden through dts with the `read-command`, `write-command`, and `rx-dummy` properties. All the read parameters are stored in RAM, so they can be overridden or completely replaced with data provided by some other mechanism, like the one from #88920.

anangl committed:
drivers: flash_mspi_nor: Add support for "supply-gpios" property

Add support for supplying power to the flash chip by activation of
a GPIO specified through the "supply-gpios" property. Implementation
of gpio_reset() is also slightly modified so that it is consistent
with soft_reset() and the new power_supply() and so that all these
functions can use a common routine that performs a reset recovery
delay.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

swift-tk reviewed (commented):
Thanks @anangl, while I appreciate the effort, there are a number of places that looks sketchy or I'm just confused by. 
Additionally, could we first merge https://github.com/zephyrproject-rtos/zephyr/pull/88920 and then work on this one?

anangl commented:
Rebased.

anangl commented on drivers/flash/flash_mspi_nor.c:
```diff
@@ -12,49 +12,122 @@
+
+		dev_data->xfer.cmd_length = 2;
+		dev_data->packet.cmd = ((uint16_t)cmd << 8)
+				     | cmd_extension;
```
You're right. This is mx25u specific behavior in 8s-8s-8s. I've added `pre_init` quirk to cover that.

anangl commented:
> Additionally, could we first merge #88920 and then work on this one?
@swift-tk With this PR, I actually wanted to limit as much as possible the stuff that needs to be defined separately for each flash chip family in #88920 (most of the commands in `flash_mspi_nor_cmds` arrays are the same there).

swift-tk commented:
> > Additionally, could we first merge #88920 and then work on this one?
> 
> @swift-tk With this PR, I actually wanted to limit as much as possible the stuff that needs to be defined separately for each flash chip family in #88920 (most of the commands in `flash_mspi_nor_cmds` structures is repeated there).

Yes, it is not quite perfect yet. The goal is to have all the common commands in flash_mspi_nor_generic (namely the core) and quirk commands override common via quirks , during or preinit, identify stage through bedecked id or SFDP.
It now provides the bare bone structure on how should the difference between flash chip is to be handled.
Though I haven’t completely finished going through this PR, I feel there might still be some conflicts. @danieldegrasse Need your help here to see if the changes are all compatible with #88920.

swift-tk reviewed (commented):
My earlier statement about SFDP still stands provided that the architecture in #88920 is generally accepted.

> The SFDP should be used for generic flash that does support it and enable through a Kconfig and dts property. With SFDP, it should load cmd and other properties through DWORDS instead of using the generic or the vendor table. It is essentially a different way of runtime probing, so I think it could coexist well provided that it doesn't get in and mess up the core all over the place.

The SFDP related configuration should be loaded in discovery stage and separated from the core driver. and enabled through a Kconfig. The configurations include context of `quad_enable_set`, `octal_enable_set`, loading of jesd216_erase_type and related functions, and more in the future. Essentially modularizing the operation of SFDP defined in jesd216.

Another bit of advise is to avoid increasingly defined RAM variables as the support of SFDP deepens: Since we would either stores DTS SFDP string or read from device in DWORDS, the config in interest could be retrieved in the form of 
```
union
{
      uint32_t DWORD_x;
      struct {
                  // bit fields defined in jesd216G
      } DWORD_xb;
}
```
Therefore enabling override to below struct(similar to `flash_mspi_nor_cmd_info`) in SFDP/QUIRK specific APIs.
```
from #88920 
struct flash_mspi_nor_cmds {
	struct flash_mspi_nor_cmd id;
	struct flash_mspi_nor_cmd write_en;
	struct flash_mspi_nor_cmd read;
	struct flash_mspi_nor_cmd status;
	struct flash_mspi_nor_cmd config;
	struct flash_mspi_nor_cmd page_program;
	struct flash_mspi_nor_cmd sector_erase;
	struct flash_mspi_nor_cmd chip_erase;
	struct flash_mspi_nor_cmd sfdp;
};
```

anangl commented on drivers/flash/flash_mspi_nor.c:
```diff
@@ -110,137 +196,150 @@ static inline uint32_t dev_flash_size(const struct device *dev)
 	} else {
-		rc = dev_cfg_apply(dev, &dev_config->mspi_nor_cfg);
+		dev_data->xfer.rx_dummy = dev_data->cmd_info.read_mode_clocks
+					+ dev_data->cmd_info.read_dummy_clocks;
```
I'll add such comment.

swift-tk commented on drivers/flash/flash_mspi_nor.c:
```diff
@@ -29,32 +42,94 @@ void flash_mspi_command_set(const struct device *dev, const struct flash_mspi_no
+		if (dev_data->in_target_io_mode) {
+			if (dev_data->packet.num_bytes != 0 ||
+			    (dev_data->xfer.addr_length != 0 &&
+			     !dev_config->single_io_addr)) {
```
I thought this is for 1-x-x modes to send cmds or register access in 1-1-1 mode which are not available in 1-x-x target io_mode? 
Here you are saying 1-1-x modes need not switch to 1-1-1 for sending cmds that are not available in 1-1-x modes which I don't think is correct.

anangl replied:
It is for 1-x-x modes. But if that non-memory-access command does not need the address phase or if the address phase in the target IO mode is configured to single line (`dev_config->single_io_addr`), there is no need to switch the IO mode for that command. That's why the `(dev_data->xfer.addr_length != 0 && !dev_config->single_io_addr)` condition is checked here. IO mode switching is omitted if it is evaluated to false, so if `(dev_data->xfer.addr_length == 0 || dev_config->single_io_addr)`, provided that the command does not need the data phase.

swift-tk commented on drivers/flash/flash_mspi_nor_sfdp.h:
```diff
@@ -0,0 +1,352 @@
+		0, \
+	.read_dummy_clocks = DT_INST_PROP_OR(inst, rx_dummy, \
+		USES_8D_8D_8D(inst) ? 20 : \
+		USES_8S_8S_8S(inst) ? 20 : \
```
Could be read from FF05 table DW6(Default Dummy Cycles After 
POR)?

anangl committed:
drivers: flash_mspi_nor: Add support for DDR in mx25u family

When Octal IO mode is to be used with DDR in mx25u family chips,
bit 1 instead of 0 must be set in the Configuration Register 2 at
address 0.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented on drivers/flash/flash_mspi_nor.c:
```diff
@@ -110,137 +196,150 @@ static inline uint32_t dev_flash_size(const struct device *dev)
-	if (dev_config->jedec_cmds->write_en.force_single) {
-		rc = dev_cfg_apply(dev, &dev_config->mspi_nor_init_cfg);
+	set_up_xfer_with_addr(dev, MSPI_RX, addr);
+	if (dev_config->dts_rx_dummy != INVALID_DTS_RX_DUMMY) {
```
Renamed this symbol to `DTS_RX_DUMMY_NOT_SPECIFIED`. Hopefully, it will be less confusing now.

anangl replied:
> okay, I see the fallback is INVALID_DTS_RX_DUMMY not 0.

No, the fallback is 0.

masz-nordic reviewed (dismissed):
A few nits, can be followed up in a separate PR.

swift-tk commented on drivers/flash/flash_mspi_nor.c:
```diff
@@ -516,198 +610,404 @@ static int dev_pm_action_cb(const struct device *dev,
+	dev_data->packet.address   = 0x02;
+	dev_data->packet.num_bytes = sizeof(uint8_t);
+	dev_data->packet.data_buf  = &status_reg;
+	rc = perform_xfer(dev, op_code, false);
```
Can we kind coding consistent? Is there any reason why cmd_rdsr can't be used here?
Is there a need to set rx_dummy and addr length here?
I think the dev_data->xfer is starting to lose sync as default in DEFAULT_CMD_INFO is not always loaded.

swift-tk commented:
@danieldegrasse Any input? Do you feel comfortable if this is merged before https://github.com/zephyrproject-rtos/zephyr/pull/88920?

anangl commented on drivers/flash/flash_mspi_nor_sfdp.h:
```diff
@@ -0,0 +1,410 @@
+
+#define DEFAULT_SWITCH_INFO(inst) { \
+	.quad_enable_req = DT_INST_ENUM_IDX(inst, quad_enable_requirements), \
+	.octal_enable_req = OCTAL_ENABLE_REQ_NONE, \
```
Yes, there is no `octal-enable-requirements` property and I don't want to introduce such thing. If needed, this field can be modified in the `pre_init` quirk for a given flash chip family.

swift-tk replied:
I think both should be added, otherwise it is SFDP only and should be wrapped with CONFIG_FLASH_MSPI_NOR_USE_SFDP or JESD216 kconfig.

anangl replied:
I'm not sure I understand the second part of your sentence. The above code is used when `CONFIG_FLASH_MSPI_NOR_USE_SFDP` is not enabled (otherwise, `quad-enable-requirements` is ignored).
I corrected it to use `DT_INST_ENUM_IDX_OR`, so that `quad-enable-requirements` is not required: https://github.com/zephyrproject-rtos/zephyr/blob/b293310a867bbcad8eefe92c673ff727c015ad1c/drivers/flash/flash_mspi_nor_sfdp.h#L410-L411

anangl replied:
No, they don't require SFDP. They still can be used when `CONFIG_FLASH_MSPI_NOR_USE_SFDP` is not enabled (the default contents of the `switch_info` structure can be overridden in `pre_init` quirk for certain flash chips.

swift-tk commented on drivers/flash/flash_mspi_nor.c:
```diff
@@ -718,6 +1018,26 @@ static int flash_chip_init(const struct device *dev)
+
+	dev_data->in_target_io_mode = true;
+
+	rc = sfdp_read(dev, 0, &sfdp_signature, sizeof(sfdp_signature));
```
> This is for testing the communication with the flash after switching to target IO mode. No SFDP data is being read here, only the signature. It's done instead of reading the JEDEC ID (which I wanted to make optional but I actually forgot to add such change in this PR), so it should not be done conditionally.

Sure, reading jedec ID can be optional. But still only jedec 216 compliant device would have the signature. I would like it maintain maximum compability. So how about wrapped it with `CONFIG_FLASH_MSPI_NOR_USE_SFDP` || 'CONFIG_FLASH_JESD216_API'?

anangl replied:
done

anangl commented:
Rebased.

maass-hamburg commented on dts/bindings/mtd/jedec,mspi-nor.yaml:
```diff
@@ -30,3 +32,17 @@ properties:
+  use-4byte-addressing:
+    type: boolean
+    description: |
+      Indicates that 4-byte addressing is to be used in communication with
+      the flash chip. The driver will use dedicated instruction codes for
+      commands that require addresses (like Read, Page Program, or Erase)
+      and will switch the flash chip to 4-byte addressing mode if necessary.
+
```
switching the flash chip to 4-byte addressing mode, is only necessary when the regular instruction codes and not the 4-byte specific ones are used. Some SoCs and FPGAs still rely that the regular instruction codes are for 3 byte addresses.

There is also already a `use-4b-addr-opcodes` prop for that in `jedec,spi-nor-common.yaml`, that this includes.

anangl replied:
I modified the `ENTER_4BYTE_ADDR()` macro so that now no switching is done if the chip supports the special 4-byte address instruction codes: https://github.com/zephyrproject-rtos/zephyr/blob/6af1921305baa350882592683346f5f08ddcea65/drivers/flash/flash_mspi_nor_sfdp.h#L302-L306
I also rephrased the above description. It should now more accurately tell what the driver does.

ExaltZephyr commented:
@anangl 

First of all, thank you for this PR — it was very helpful for me when testing my flash with 4-byte addressing mode.

I do have one concern: I noticed that you removed all flash_mspi_nor_cmds definitions from the header. Previously, the JEDEC driver used the force_single property in flash_mspi_nor_cmds to determine whether a command had to be executed in single mode. Some commands must always be issued in single mode, so this check was important to apply the correct configuration before performing read/write/erase operations.

Could you clarify how this requirement is handled with your changes?

anangl commented:
@ExaltZephyr
> I do have one concern: I noticed that you removed all flash_mspi_nor_cmds definitions from the header. Previously, the JEDEC driver used the force_single property in flash_mspi_nor_cmds to determine whether a command had to be executed in single mode. Some commands must always be issued in single mode, so this check was important to apply the correct configuration before performing read/write/erase operations.
> 
> Could you please clarify how this requirement is handled with your changes?

Switching to Single IO mode is done if needed for commands other than Read and Page Program. The relevant part of code is [here](https://github.com/zephyrproject-rtos/zephyr/blob/de71be41813fb0089e84b320b680042897348b89/drivers/flash/flash_mspi_nor.c#L84-L123).

SaraATouqan commented:
> @ExaltZephyr
> 
> > I do have one concern: I noticed that you removed all flash_mspi_nor_cmds definitions from the header. Previously, the JEDEC driver used the force_single property in flash_mspi_nor_cmds to determine whether a command had to be executed in single mode. Some commands must always be issued in single mode, so this check was important to apply the correct configuration before performing read/write/erase operations.
> > Could you please clarify how this requirement is handled with your changes?
> 
> Switching to Single IO mode is done if needed for commands other than Read and Page Program. The relevant part of code is [here](https://github.com/zephyrproject-rtos/zephyr/blob/de71be41813fb0089e84b320b680042897348b89/drivers/flash/flash_mspi_nor.c#L84-L123).

@anangl In my case, the code always enters the first condition "if (dev_config->multi_io_cmd ...)" since I am using octal mode, and this seems to always evaluate to true regardless of the command opcode. Am I missing any configuration here?

anangl commented:
Rebased.

anangl commented:
> @anangl In my case, the code always enters the first condition "if (dev_config->multi_io_cmd ...)" since I am using octal mode, and this seems to always evaluate to true regardless of the command opcode. Am I missing any configuration here?

@ExaltZephyr If you're using octal mode, then switching is not needed, so indeed the code should enter the first branch.

anangl commented:
> Switching to single mode is required for certain commands, as some JEDEC commands must be executed in single mode even when operating in dual, quad, or octal modes.

Can you provide some examples for which the new code would not work properly?

> @ExaltZephyr If you're using octal mode, then switching is not needed, so indeed the code should enter the first branch.

To clarify, by "octal mode" I only mean `MSPI_IO_MODE_OCTAL`, not `*_1_1_8` or `*_1_8_8`, for which `multi_io_cmd` will be `false`, so switching may be performed for them.

anangl committed:
samples: code_relocation_nocopy: Update configuration for nRF54H20 DK

It is no longer needed to specify the read command parameters in dts,
as the flash driver is capable of obtaining those from SFDP structures,
so the redundant properties are removed.

After commit 612fd940fa307ea8a83b97571505d6201ab28d85 got merged, both
the EXMIF and flash nodes in dts are not enabled by default so this
needs to be done in the overlay.

Although nrfutil is still not capable of programming the external
flash on nRF54H20 DK, thus `west flash` will not work in this case,
`nrf54h20dk/nrf54h20/cpuapp` is added as an allowed platform so that
the sample is at least built in CI for it.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

swift-tk commented on drivers/flash/Kconfig.mspi:
```diff
@@ -45,10 +45,29 @@ menuconfig FLASH_MSPI_NOR
 
+config FLASH_MSPI_NOR_USE_SFDP
+	bool "Use Serial Flash Discoverable Parameters (SFDP)"
+	default y
```
I think it might be a good idea to add depends on DT_HAS_JEDEC_JESD216_ENABLED. But this will need to add the compat string in binding and DTS.... 
If the above is not feasible, I'm also okay with some other checking that uses the presence of sfdp related property in DTS, so that it is not always enabled by default even if there is no `sfdp-bfp` and etc.

maass-hamburg commented on dts/bindings/mtd/jedec,mspi-nor.yaml:
```diff
@@ -30,3 +45,18 @@ properties:
       bus in communication with the flash chip. The default value is the one
       that was previously hard-coded in the flash_mspi_nor driver.
+
+  use-4byte-addressing:
```
wouldn't it make sense, to automatically enable this, when the size of the flash is to big to use 3 byte addressing.

maass-hamburg replied:
Buth then we could do it like in the regular spi_nor, where the address is checked if it fits in 3 bytes and if it does 3 byte addressing is used and only for addresses that don't fit 4 byte mode is used. Could then still have a option to always force 4 byte mode

carlescufi replied:
@maass-hamburg sounds like a potential improvement, perhaps not enough to block this PR at this point?

ExaltZephyr commented:
@anangl  Hi
Were you able to run your flash in MSPI_IO_OCTAL_MODE without any issues when using the status command ?

Also, I noticed that you added cmd_extension only for Octal DTR mode.
However, we also need it for Octal SDR since its opcodes are 2 bytes. Am I missing something here?

<img width="876" height="165" alt="image" src="https://github.com/user-attachments/assets/26aed8c1-eadb-48f2-a860-4bc9dadb0b5f" />
